### ISSUE TITLE
ENH: GFDL options for custom weights and activation functions

### ISSUE THREAD

**Issue by aganguly-lab** [lanl/GFDL#124]
As mentioned in the title, I think it may be useful to include an option for users to apply custom-built activation functions or weighting schemes. This will be useful for using the GFDL library to test Kyle's theoretical predictions (the current GFDL library does not accommodate the scalings that Kyle uses, and I'm not sure if we want to hard-code those kinds of scalings in or add more parameters to make scaling optional).

It will also be useful in future testing to have some method of implementing custom weighting schemes, particularly non-i.i.d. weighting schemes (e.g. importing weights from trained back-propagated neural networks). 

I'm curious to hear your feedback on whether this would also be something worth adding to the scikit-learn version of the project, or if what we have submitted there is sufficient.

**Comment by tylerjereddy** [comment]
Any literature evidence of these less common scaling schemes being used in well cited works outside of research experiments on our team? I thought we covered the main ones we saw in the literature already.

I'm not really sure why the theoretical side keeps messing around with internal scaling given that it seems to be rare in practical usage/implementations. Would be good to have a motivation other than theoreticians are using scaling for convenience or something like that. Why are theoreticians using scaling so much? I still haven't seen a clear explanation rooted in practical day to day usage.

Have you checked that we for sure can't simply access the weights/details of the models as `_private` or `public_` attributes of the estimators and set them that way to override things in these specific research scenarios?

**Comment by aganguly-lab** [comment]
I can check the literature to see if there is any analysis of what, if anything, goes wrong without the scaling. @nray and I are currently experimenting with both including and removing the scalings from GFDL, so hopefully we will soon have a better idea of the empirical affect of scaling on GFDL. 

As for your question about accessing internal details of the model, how would that work? Would we make the minimal change to our local source code and go from there? If so, I think we can achieve Kyle's scaling's by adding an option to multiply the activation functions by a constant and by the reciprocal square root of `self.hidden_layer_size`. There are some non-standard activation functions we also plan to test so we can hopefully understand what is happening (sign function as an approximation of what happens when there is no d scaling and n is large, cube root function to see if d scaling performance improves for functions with an infinite derivative at 0). With these minimal modifications, it should be possible to use GFDL with the theoretical scalings by adding a few extra pre-processing and post-processing steps. I just thought it would be more convenient to make all of the changes in one place (this PR) to take advantage of the existing test suite and just add a few additional tests. But I suppose we could just add a few additional tests to @nray's MR instead and simply make sure our local GFDL changes don't break the project.

However, my thoughts for this issue were more long-term. @sdtemple and I have been talking about testing some methods using non-i.i.d. weights which may be able to improve the performance of GFDL. I think this is a direction where empirical studies should precede the theory because non-i.i.d. analysis is difficult without some guiding intuition to rely on. I also think that this will be an important direction of exploration as we advance the project, so it would be nice if we could have a well-tested method of adding custom weights/activation functions before we need it. I recognize the primary use of this change would be for research, so I don't think we should add it to the scikit-learn PR.

**Comment by tylerjereddy** [comment]
Probably the first step in convincing me is to write quality regression tests showing that the various scaling approaches you want to support are not generally deleterious to performance on common classification/regression datasets. Even that is a bit weak--if they are only roughly in the same ballpark as "no scaling" one could easily ask--why all this extra complexity for the roughly the same performance? I'll try to take a look at the PR when I can. I believe you've done a little testing on the standard MNIST digits dataset there, but I'd probably look for more than that.

> I recognize the primary use of this change would be for research, so I no longer believe we should add it to the scikit-learn PR.

Yes, it is completely out of scope for `sklearn` unless there is strong evidence of superior performance with scaling. Even for this project, the review process should be quite rigorous, but we do have a bit more of an angle toward research in this project, if it makes sense. Adding bells and whistles has a complexity/maintenance cost even here though, and keeping our tests passing even with the relatively simple infrastructure we currently have is already a challenge.

If you want to support arbitrary user-supplied activations/weights/whatever, then to some degree the burden of testing those custom functions may fall back to Navamita's MR/wherever you have the custom functions. If you want a finite set of sensible new functionality here, then yeah maybe we can carefully test it here.

**Comment by tylerjereddy** [comment]
> As for your question about accessing internal details of the model, how would that work? 

Probably just a hack like `model._weights = your_custom_weights` or something like that. It isn't very satisfying to do that without accompanying tests/official infrastructure, but if the outcome of the review here is that this a bunch of complexity for not much benefit it might be a last resort. Actually, it might not be that bad either, since copying the weight schemes of pre-trained models directly in may be desirable at some point anyway, so the ability to patch over the random weights may be relevant..

**Comment by aganguly-lab** [comment]
> Probably the first step in convincing me is to write quality regression tests showing that the various scaling approaches you want to support are not generally deleterious to performance on common classification/regression datasets. 

We will test and get back to you. The purpose of PR #125 was to quickly provide an environment to perform those tests where we can be reasonably confident that any results we find are empirically valid rather than caused by uncaught bugs.

>Probably just a hack like `model._weights = your_custom_weights` or something like that. It isn't very satisfying to do that without accompanying tests/official infrastructure, but if the outcome of the review here is that this a bunch of complexity for not much benefit it might be a last resort.

This will work for ad-hoc testing (testing ideas with outcomes we are uncertain of). But I think it will be a good idea to, at the very least, develop a sanctioned method of inputting custom weights, such as weights from pretrained models or any non-i.i.d. weights which can be demonstrated to significantly improve the performance of GFDL. Perhaps it could be as simple as adding a 'custom' option to the `weight_scheme` input, and if this argument is called, then instead of building the weights as we already do, the code simply imports a weights_ array from a user-provided file and performs the relevant checks to make sure the imported weights are valid (and returns an error if not)? The weight construction process is already very nicely localized in one place each for ensemble and non-ensemble methods.

As for testing, I suppose an alternative possibility might be to create a separate local library populated by different experimental methods of generating weights/activation functions, importing it in `model.py`, adding `model._weights = custom_weights([input])`, then indexing our experiments by the `custom_weights` method used (in addition to any other parameters). We can probably do something similar if we have any reason to test weird activation functions. We'd just need to be careful during testing that random seed behavior is consistent between runs.
